File: docs/examples/cpp/src/data_querying_filter.cc
#include <reduct/client.h>
#include <iostream>
#include <cassert>

using reduct::IBucket;
using reduct::IClient;
using reduct::Error;

using std::chrono_literals::operator ""s;

int main() {
    // Create a client instance, then get or create a bucket
    auto client = IClient::Build("http://127.0.0.1:8383", {.api_token="my-token"});
    auto [bucket, create_err] = client->GetOrCreateBucket("example-bucket");
    assert(create_err == Error::kOk);

    // Query 10 photos from "imdb" entry which taken after 2006 with the face score less than 4
    auto err = bucket->Query("imdb", std::nullopt, std::nullopt, {
            .when=R"({
                "&photo_taken": {"$gt": 2006},
                "&name": {"$lt": 4}
            })",
            .limit = 10,
    }, [](auto rec) {
        std::cout << "Name: " << rec.labels["name"] << std::endl;
        std::cout << "Photo Taken: " << rec.labels["photo_taken"] << std::endl;
        std::cout << "Face Score: " << rec.labels["face_score"] << std::endl;

        auto [_, read_err] = rec.ReadAll();
        assert(read_err == Error::kOk);
        return true;    // if false, the query will stop
    });

    assert(err == Error::kOk);
    return 0;
}

File: versioned_docs/version-1.14.x/examples/rs/examples/data_querying_filter.rs
use futures::StreamExt;
use reduct_rs::{condition, ReductClient, ReductError};
use tokio;

#[tokio::main]
async fn main() -> Result<(), ReductError> {
    // Create a client instance, then get or create a bucket
    let client = ReductClient::builder()
        .url("http://127.0.0.1:8383")
        .api_token("my-token")
        .build();
    let bucket = client
        .create_bucket("example-bucket")
        .exist_ok(true)
        .send()
        .await?;

    // Query 10 photos from "imdb" entry which taken after 2006 with the face score less than 4
    let query = bucket
        .query("imdb")
        .when(condition!({
            "&photo_taken": {"$gt": 2006},
            "&face_score": {"$gt": 4}
        }))
        .limit(10)
        .send()
        .await?;

    tokio::pin!(query);
    while let Some(record) = query.next().await {
        let record = record?;
        println!("Name: {:?}", record.labels().get("name"));
        println!("Photo Taken: {:?}", record.labels().get("photo_taken"));
        println!("Face Score: {:?}", record.labels().get("face_score"));

        _ = record.bytes().await?;
    }

    Ok(())
}
